addons/api/addon/utils/indexedDbQuery.js

```diff
@@ -0,0 +1,276 @@
+    case 'gte':
+      return baseWhereClause.aboveOrEqual(value);
+    case 'lt':
+      return baseWhereClause.below(value);
+    case 'lte':
+      return baseWhereClause.belowOrEqual(value);
+    case 'notEquals':
+      return baseWhereClause.notEqual(value);
+    default:
+      return baseWhereClause.equals(value);
+  }
+};
+
+/**
+ * Continues building the queries for indexed DB to filter on. Note that these are not true
+ * indexed DB queries and in fact just standard JS filters for logical "and" queries and parallel queries
+ * to indexed DB for logical "or" queries that get deduped.
+ */
+const buildIndexedDbCollection = ({
+  key,
+  value,
+  collection,
+  operation,
+  logicalOperator,
+}) => {
+  switch (operation) {
+    case 'contains':
+      if (logicalOperator === 'and') {
+        return collection.and((record) => get(record, key)?.includes(value));
+      } else {
+        return collection
+          .or(key)
+          .filter((record) => get(record, key)?.includes(value));
+      }
+    case 'gt':
+      if (logicalOperator === 'and') {
+        return collection.and((record) => get(record, key) > value);
+      } else {
+        return collection.or(key).above(value);
+      }
+    case 'gte':
+      if (logicalOperator === 'and') {
+        return collection.and((record) => get(record, key) >= value);
+      } else {
+        return collection.or(key).aboveOrEqual(value);
+      }
+    case 'lt':
+      if (logicalOperator === 'and') {
+        return collection.and((record) => get(record, key) < value);
+      } else {
+        return collection.or(key).below(value);
+      }
+    case 'lte':
+      if (logicalOperator === 'and') {
+        return collection.and((record) => get(record, key) <= value);
+      } else {
+        return collection.or(key).belowOrEqual(value);
+      }
+    case 'notEquals':
+      if (logicalOperator === 'and') {
+        return collection.and((record) => get(record, key) !== value);
+      } else {
+        return collection.or(key).notEqual(value);
+      }
+    default:
+      if (logicalOperator === 'and') {
+        return collection.and((record) => get(record, key) === value);
+      } else {
+        return collection.or(key).equals(value);
+      }
+  }
+};
+
+/**
+ * Translates the fields passed in from the UI for resources to the keys indexed DB knows them as.
+ * Usually it's the path of the actual normalized model.
+ */
+const getKey = (key) => {
+  switch (key) {
+    case 'scope_id':
+      return `attributes.scope.${key}`;
+    case 'id':
+      return key;
+    default:
+      return `attributes.${key}`;
+  }
+};
+
+/**
+ * Cleans the object to remove any keys that have a value of an empty array. This makes
+ * it easier so we don't have to worry about empty arrays getting queried in indexed db.
+ */
+const cleanObject = (obj) => {
+  if (!obj) {
+    return {};
+  }
+
+  return Object.entries(obj).reduce((result, [key, value]) => {
```

Comment:
curious why you didn't go for a `for.. in` and just simply delete the keys with empty array 
something like 
`for (const key in obj) { if (Array.isArray(obj[key]) && !obj[key].length) delete obj[key] }`

Reply:
Mainly because I didn't want to modify the array and prefer to return a new one instead. I like to reduce the side effects in these type of helper functions